source/agora/test/EnrollmentManager.d

```diff
@@ -142,3 +142,82 @@ unittest
     assert(b20.header.enrollments.length == 1);
     assert(b20.header.enrollments[0] == enroll);
 }
+
+/// Situation: A node in the network crashes after a new enrollment for the
+///     node has been inserted into a new block with consensus. And the node
+///     starts again
+/// Expectation: The node restores its own enrollment from the chain and
+///     reveals its pre-images preidically.
+unittest
+{
+    // Boilerplate
+    TestConf conf = {
+        validator_cycle : 20,
+    };
+    auto network = makeTestNetwork(conf);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+
+    // Sanity check: Check if genesis block has enrollments
+    const b0 = nodes[0].getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length >= 1);
+
+    // Make 19 blocks
+    Transaction[] txs;
+    foreach (count; 0 .. 19)
+    {
+        // Use Genesis
+        if (!txs.length)
+            txs = genesisSpendable().take(Block.TxsInBlock).enumerate()
+                .map!(en => en.value.refund(WK.Keys.A.address).sign())
+                .array();
+        else
+            txs = txs
+                .map!(ptx => TxBuilder(ptx).refund(WK.Keys[count].address).sign())
+                .array();
+        txs.each!(tx => nodes[1].putTransaction(tx));
+
+        // Ensure everyone is at the same level
+        ensureConsistency(nodes, count + 1);
+    }
+
+    // Now create an Enrollment for nodes[0], create block #20, and restart nodes[0]
+    auto enroll = nodes[0].createEnrollmentData();
+    nodes[0].enrollValidator(enroll);
+    txs = txs
+        .map!(ptx => TxBuilder(ptx).refund(WK.Keys[20].address).sign())
+        .array();
+    txs.each!(tx => nodes[1].putTransaction(tx));
+    ensureConsistency(nodes.take(1), 20);
+    assert(nodes[0].getBlocksFrom(20, 1)[0].header.enrollments.length == 1);
+
+    network.restart(nodes[0]);
+    network.waitForDiscovery();
+    ensureConsistency([nodes[0]], 20, 10.seconds);
+
+    // Now make a new block and make sure only nodes[1] signs it
```

Comment:
```suggestion
    // Now make a new block and make sure only nodes[0] signs it
```

Right?

Reply:
>     // Now make a new block and make sure only nodes[0] signs it
> Right?

The nodes[1] is right. We used that intentionally in the pair-programming with @Geod24

Reply:
Is the comment referring to signing the transaction or the block? That's what I was confused by the wording. "make a new block and sign it" refers to the block.

Reply:
By the way, If I change the code, does the code show the situation that only nodes[0] sign it?

Reply:
Sign what? That was my question.

Reply:
It was written by @Geod24 so I think he intended `block`.

Reply:
Hmm.. But I think all of the node's enrollments have expired at block height 20, except node[0] which re-enrolled. So I'm not sure that comment was correct? Let's see what he says.

Reply:
Anyway there is no actual verification here of **which node** created the block. So "make sure" doesn't make sense here.

Reply:
OK, I will modify it to the right expression after checking.

Reply:
At last, I agree with you. It should be `nodes[0]`. Sorry for my late understanding. Thanks.